draco/asp/soft.lp

```diff
@@ -30,18 +30,67 @@ preference(same_field,F) :-
     entity(mark,_,M),
     { entity(encoding,M,E): attribute((encoding,field),E,F) } = 2.
 
-% @soft(type_linear) linear scale
-preference(type_linear,E) :-
+% @soft(same_field_grt3) Prefer not to use the same field three or more times for the same mark.
+preference(same_field_grt3,F) :-
+    entity(field,_,F),
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,field),E,F) } >= 3.
+
+% @soft(count_grt1) Prefer not to use count more than once.
+preference(count_grt1,M) :-
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,aggregate),E,count) } > 1.
+
+% @soft(shape_high_cardinality) Shape channel should not have too high cardinality.
+preference(shape_high_cardinality,E) :-
+    attribute((encoding,channel),E,shape),
+    attribute(encoding_cardinality,E,N),
+    N > 5.
+
+% @soft(number_categorical) Numbers should not be nominal.
+preference(number_categorical,F) :-
+    attribute((field,type),F,number),
+    attribute(field_scale,F,categorical).
+
+% @soft(bin_number_low_cardinality) Binned number field should not have too low cardinality.
+preference(bin_number_low_cardinality,E) :-
+    attribute((field,type),F,number),
+    attribute((encoding,field),E,F),
+    attribute((encoding,binning),E,B),
+    B < 15.
+
+% @soft(bin_not_linear) Prefer linear scale for bin.
+preference(bin_not_linear,E) :-
+    attribute((encoding,binning),E,_),
+    not attribute((encoding,scale_type),E,linear).
+
+% @soft(only_discrete) Plots with only nominal, ordinal, binned q, or t with time unit should add either an aggregation (e.g. count) or a quantitative field.
```

Comment:
The constraint doesn't match the description here. I think the description says that every encoding of the plot is discrete, but here you are only saying the encodings for one mark are discrete.

Reply:
changed discription